File: R/twliConsolidate/R/twli.R
#' Sum specified columns within each consolidation group (對應 ``liSum``).
#'
#' Summed values are written onto the keeper row (``merge_code == 1``); other
#' rows are left untouched. When ``finish=TRUE`` the non-keeper rows are
#' dropped and an ``li_adjust`` column (1 = aggregated keeper, 0 = unchanged)
#' is appended.
#'
#' @param x A data.frame with at least ``vid_col`` and the columns named in
#'   ``cols``.
#' @param ref Reference table from [liRef()].
#' @param cols Character vector of columns to sum.
#' @param finish Logical; if TRUE, drop non-keeper rows and add
#'   ``li_adjust``.
#' @param vid_col Name of the village-code column. Default ``"V_ID"``.
#' @export
liSum <- function(x, ref, cols, finish = FALSE, vid_col = "V_ID") {
  .validate(x, ref, cols, vid_col)
  out <- x

  joined <- merge(out[, c(vid_col, cols), drop = FALSE], ref,
                  by.x = vid_col, by.y = "v_id")
  sums <- stats::aggregate(joined[, cols, drop = FALSE],
                           by = list(group = joined$group),
                           FUN = function(v) sum(v, na.rm = TRUE))
  keepers <- ref[ref$merge_code == 1L, c("v_id", "group")]
  rownames(sums) <- as.character(sums$group)
  for (i in seq_len(nrow(keepers))) {
    v <- keepers$v_id[i]; g <- as.character(keepers$group[i])
    rowi <- which(out[[vid_col]] == v)
    if (length(rowi)) for (c in cols) out[rowi, c] <- sums[g, c]
  }
  .maybe_finalise(out, ref, vid_col, finish)
}


#' Recompute a column from a formula after group-summing the operands
#' (對應 ``liEqu``).
#'
#' @param x A data.frame.
#' @param ref Reference table from [liRef()].
#' @param result Name of the column to write the recomputed value to.
#' @param equ Character expression, e.g. ``"P_CNT / AREA"``. Operand columns
#'   are auto-detected, summed, then ``equ`` is evaluated on the sums.
#' @param finish Logical; same as in [liSum()].
#' @param vid_col Name of the village-code column.
#' @section Convention:
#'   Run all ``liEqu`` calls first on the raw data **before** running
#'   ``liSum`` on the same operand columns; otherwise the operands will be
#'   double-counted.
#' @export
liEqu <- function(x, ref, result, equ, finish = FALSE, vid_col = "V_ID") {
  operand_cols <- .extract_columns(equ, names(x))
  .validate(x, ref, operand_cols, vid_col)
  out <- x

  joined <- merge(out[, c(vid_col, operand_cols), drop = FALSE], ref,
                  by.x = vid_col, by.y = "v_id")
  sums <- stats::aggregate(joined[, operand_cols, drop = FALSE],
                           by = list(group = joined$group),
                           FUN = function(v) sum(v, na.rm = TRUE))
  sums[[result]] <- eval(parse(text = equ), envir = sums)

  if (!result %in% names(out)) out[[result]] <- NA_real_

  keepers <- ref[ref$merge_code == 1L, c("v_id", "group")]
  rownames(sums) <- as.character(sums$group)
  for (i in seq_len(nrow(keepers))) {
    v <- keepers$v_id[i]; g <- as.character(keepers$group[i])
    rowi <- which(out[[vid_col]] == v)
    if (length(rowi)) out[rowi, result] <- sums[g, result]
  }
  .maybe_finalise(out, ref, vid_col, finish)
}


#' Geometric union of polygons within each consolidation group
#' (對應 ``liShp``).
#'
#' Uses ``sf::st_union`` to combine geometries within each group, replaces the
#' keeper row's geometry with the union, then drops the non-keeper rows. An
#' ``li_adjust`` column is added.
#'
#' @param x An ``sf`` object with ``vid_col``.
#' @param ref Reference table from [liRef()].
#' @param vid_col Name of the village-code column.
#' @export
liShp <- function(x, ref, vid_col = "V_ID") {
  if (!requireNamespace("sf", quietly = TRUE))
    stop("liShp requires the 'sf' package")
  if (!inherits(x, "sf"))
    stop("liShp expects an 'sf' object")

  out <- x
  out$`__group__` <- ref$group[match(out[[vid_col]], ref$v_id)]

  # Compute union per group
  in_groups <- out[!is.na(out$`__group__`), ]
  if (nrow(in_groups) > 0) {
    union_per_group <- list()
    for (g in unique(in_groups$`__group__`)) {
      union_per_group[[as.character(g)]] <-
        sf::st_union(sf::st_geometry(in_groups[in_groups$`__group__` == g, ]))
    }
    keepers <- ref[ref$merge_code == 1L, ]
    for (i in seq_len(nrow(keepers))) {
      v <- keepers$v_id[i]; g <- as.character(keepers$group[i])
      idx <- which(out[[vid_col]] == v)
      if (length(idx) && !is.null(union_per_group[[g]])) {
        sf::st_geometry(out)[[idx[1]]] <- union_per_group[[g]][[1]]
      }
    }
  }

  non_keeper_v <- ref$v_id[ref$merge_code != 1L]
  out <- out[!out[[vid_col]] %in% non_keeper_v, ]
  out$li_adjust <- ifelse(out[[vid_col]] %in% ref$v_id[ref$merge_code == 1L], 1L, 0L)
  out$`__group__` <- NULL
  out
}


# ---------- internals ----------

.NUM_RE <- "^[0-9.]+$"

.extract_columns <- function(equ, available) {
  toks <- strsplit(gsub("\\s+", " ", equ), "[\\s/+\\-*^()(),]+", perl = TRUE)[[1]]
  toks <- toks[nchar(toks) > 0 & !grepl(.NUM_RE, toks)]
  out <- intersect(unique(toks), available)
  if (!length(out))
    stop(sprintf("No known columns referenced in expression %s", equ))
  out
}

.validate <- function(x, ref, cols, vid_col) {
  if (!vid_col %in% names(x)) stop(sprintf("vid_col '%s' not in data", vid_col))
  for (c in c("v_id", "group", "merge_code")) {
    if (!c %in% names(ref)) stop(sprintf("ref missing column '%s'", c))
  }
  miss <- setdiff(cols, names(x))
  if (length(miss)) stop(sprintf("columns missing in data: %s",
                                 paste(miss, collapse = ", ")))
}

.maybe_finalise <- function(out, ref, vid_col, finish) {
  if (!finish) return(out)
  non_keeper_v <- ref$v_id[ref$merge_code != 1L]
  keeper_v <- ref$v_id[ref$merge_code == 1L]
  out <- out[!out[[vid_col]] %in% non_keeper_v, ]
  out$li_adjust <- ifelse(out[[vid_col]] %in% keeper_v, 1L, 0L)
  rownames(out) <- NULL
  out
}

File: python/twli_consolidate/twli.py
"""twli-style aggregation functions: li_sum / li_equ / li_shp.

These mirror the API of the original R package
`torrentpien/twli <https://github.com/torrentpien/twli>`_ but operate on the
Python crosswalk produced by :func:`build_ref`.
"""
from __future__ import annotations

import re
from typing import Iterable

import pandas as pd


def li_sum(
    df: pd.DataFrame,
    ref: pd.DataFrame,
    cols: Iterable[str] | str,
    finish: bool = False,
    vid_col: str = "V_ID",
) -> pd.DataFrame:
    """Sum specified columns within each consolidation group (對應 ``liSum``).

    The summed values are written onto the row whose ``merge_code == 1`` (the
    keeper). Other group-member rows are left untouched (so ``li_equ`` can be
    chained afterwards). When ``finish=True`` the non-keeper rows are dropped
    and an ``li_adjust`` column (1 = aggregated keeper, 0 = unchanged) is
    appended.
    """
    cols = [cols] if isinstance(cols, str) else list(cols)
    _validate(df, ref, cols, vid_col)

    out = df.copy()
    joined = out[[vid_col, *cols]].merge(
        ref, left_on=vid_col, right_on="v_id", how="inner"
    )
    sums = joined.groupby("group", as_index=False)[cols].sum(min_count=1)

    keepers = ref[ref["merge_code"] == 1][["v_id", "group"]]
    keeper_map = keepers.set_index("v_id")["group"].to_dict()
    sums_map = sums.set_index("group")

    keeper_rows = out[vid_col].isin(keeper_map.keys())
    for idx in out[keeper_rows].index:
        v = out.at[idx, vid_col]
        g = keeper_map[v]
        for c in cols:
            out.at[idx, c] = sums_map.at[g, c]

    return _maybe_finalise(out, ref, vid_col, finish)


def li_equ(
    df: pd.DataFrame,
    ref: pd.DataFrame,
    result: str,
    equ: str,
    finish: bool = False,
    vid_col: str = "V_ID",
) -> pd.DataFrame:
    """Recompute a column from a formula after group-summing the operands
    (對應 ``liEqu``).

    ``equ`` is a string expression such as ``"P_CNT / AREA"``. Column names
    referenced inside ``equ`` are auto-detected, summed within each group,
    then ``equ`` is evaluated on the summed values and written to the keeper
    row's ``result`` column.

    Convention
    ----------
    Call all ``li_equ`` invocations on the raw dataframe **before** running
    ``li_sum`` on the same operand columns. ``li_equ`` re-sums the operand
    columns each call, so feeding it a dataframe whose operand columns have
    already been replaced by ``li_sum`` will double-count the keeper row.
    """
    operand_cols = _extract_columns(equ, df.columns)
    _validate(df, ref, operand_cols, vid_col)

    out = df.copy()
    joined = out[[vid_col, *operand_cols]].merge(
        ref, left_on=vid_col, right_on="v_id", how="inner"
    )
    sums = joined.groupby("group", as_index=False)[operand_cols].sum(min_count=1)
    sums[result] = sums.eval(equ)

    keepers = ref[ref["merge_code"] == 1][["v_id", "group"]]
    keeper_map = keepers.set_index("v_id")["group"].to_dict()
    sums_map = sums.set_index("group")[result].to_dict()

    if result not in out.columns:
        out[result] = pd.NA

    keeper_rows = out[vid_col].isin(keeper_map.keys())
    for idx in out[keeper_rows].index:
        v = out.at[idx, vid_col]
        g = keeper_map[v]
        out.at[idx, result] = sums_map[g]

    return _maybe_finalise(out, ref, vid_col, finish)


def li_shp(
    gdf,
    ref: pd.DataFrame,
    vid_col: str = "V_ID",
):
    """Geometric union of polygons within each consolidation group
    (對應 ``liShp``).

    Replaces the keeper-row geometry with ``shapely.unary_union`` of all group
    members and drops the non-keeper rows. ``gdf`` should be a
    :class:`geopandas.GeoDataFrame`.
    """
    try:
        import geopandas as gpd  # type: ignore
        from shapely.ops import unary_union
    except ImportError as e:  # pragma: no cover
        raise ImportError(
            "li_shp requires geopandas + shapely. Install via "
            "`pip install geopandas shapely`."
        ) from e
    if not isinstance(gdf, gpd.GeoDataFrame):
        raise TypeError("li_shp expects a geopandas.GeoDataFrame")

    out = gdf.copy()
    keepers = ref[ref["merge_code"] == 1][["v_id", "group"]]
    keeper_v_to_g = keepers.set_index("v_id")["group"].to_dict()
    member_v_to_g = ref.set_index("v_id")["group"].to_dict()

    # Compute unioned geometry per group
    out["__group__"] = out[vid_col].map(member_v_to_g)
    grouped_geoms = (
        out[out["__group__"].notna()]
        .groupby("__group__")["geometry"]
        .apply(lambda s: unary_union(list(s.values)))
    )

    # Replace keeper-row geometry
    for v, g in keeper_v_to_g.items():
        mask = out[vid_col] == v
        if mask.any() and g in grouped_geoms.index:
            out.loc[mask, "geometry"] = grouped_geoms.loc[g]

    # Drop non-keeper rows
    non_keeper_v = ref[ref["merge_code"] != 1]["v_id"].tolist()
    out = out[~out[vid_col].isin(non_keeper_v)].copy()
    out["li_adjust"] = out[vid_col].map(
        lambda v: 1 if v in keeper_v_to_g else 0
    ).astype(int)
    out = out.drop(columns=["__group__"], errors="ignore").reset_index(drop=True)
    return out


# ---------- internals ----------

_NUM_RE = re.compile(r"^[0-9.]+$")


def _extract_columns(expr: str, available: Iterable[str]) -> list[str]:
    """Pull identifiers out of an arithmetic expression like ``"P_CNT / AREA"``."""
    tokens = re.split(r"[/+\-*^()\s,]+", expr)
    available = set(available)
    out: list[str] = []
    for t in tokens:
        if not t or _NUM_RE.match(t):
            continue
        if t in available and t not in out:
            out.append(t)
    if not out:
        raise ValueError(
            f"No known columns referenced in expression {expr!r}; "
            "make sure operand names match the dataframe."
        )
    return out


def _validate(df: pd.DataFrame, ref: pd.DataFrame, cols: list[str], vid_col: str) -> None:
    if vid_col not in df.columns:
        raise KeyError(f"vid_col {vid_col!r} not in dataframe")
    for c in ("v_id", "group", "merge_code"):
        if c not in ref.columns:
            raise KeyError(f"ref dataframe missing required column {c!r}")
    missing = [c for c in cols if c not in df.columns]
    if missing:
        raise KeyError(f"columns missing in dataframe: {missing}")


def _maybe_finalise(out: pd.DataFrame, ref: pd.DataFrame,
                    vid_col: str, finish: bool) -> pd.DataFrame:
    if not finish:
        return out
    non_keeper_v = set(ref.loc[ref["merge_code"] != 1, "v_id"])
    keeper_v = set(ref.loc[ref["merge_code"] == 1, "v_id"])
    out = out[~out[vid_col].isin(non_keeper_v)].copy()
    out["li_adjust"] = out[vid_col].map(lambda v: 1 if v in keeper_v else 0).astype(int)
    return out.reset_index(drop=True)
